src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueue.java

```diff
@@ -776,11 +776,13 @@ public long firstIndex() {
     public long lastIndex() {
         // This is a slow implementation that gets a Tailer/DocumentContext to find the last index
         try (final ExcerptTailer tailer = createTailer().direction(BACKWARD).toEnd()) {
-            try (final DocumentContext documentContext = tailer.readingDocument()) {
-                if (documentContext.isPresent()) {
-                    return documentContext.index();
-                } else {
-                    return -1;
+            while (true) {
+                try (final DocumentContext documentContext = tailer.readingDocument()) {
+                    if (documentContext.isPresent() && !documentContext.isMetaData()) {
+                        return documentContext.index();
+                    } else {
+                        return -1;
+                    }
```

Comment:
is the while (true) redundant here? it seems regardless of which branch you take you return something on the first iteration?

Reply:
I'm guessing in the event it's metadata we want to keep scanning backwards?

Reply:
and we should have a timeout rather than loop forever?

Reply:
It won't loop forever because it's traversing the queue backwards, it'll hit the first cycle then exit.